### PR TITLE
lexer/parser: 2.5 support

### PR DESCRIPTION
**Pull request by iliabylich**
Updated lexer.rl and typedruby.ypp to match whitequark/parser.
Changes to the test suite:
1. Added a script that imports all tests from `whitequark/parser`.
2. This script saves input/output files in `parse-tree-whitequark` format because it's a) more compact and b) matches what `whitequark/parser` prints, so it simplifies debugging.
3. Added a `//test:whitequark_parser_tests` test bazel task that runs these files with:
  a. only parsing phase
  b. `foo`, `bar`, `baz` local variables (just like `whitequark/parser`)
4. Excluded all `whitequark/parser` tests from the `test_corpus`
5. re-generated all files in `test/testdata/parser/whitequark/*`

Changes to the code:
1. Updated both `lexer.rl` and `typedruby.ypp` to match `whitequark/parser`.
2. `foo while bar` is no longer a post-while loop (`raise while false` doesn't raise, it's just a regular while loop), same for `foo unless bar`
3. defining constants/modules/dynamic constants in methods is no longer allowed
4. fixed parsing of `def BEGIN; end`
5. `lex.in_kwarg` is used now (I think typedruby removed it intentonally), so the code like `def a b:\nreturn\nend` can be parsed now.
6. removed redundant `begin`, `dstr` and `dsym` nodes from some placees. The AST is a bit more compact now.
7. introduced `ruby_parser::Context`
8. fixed parsing of multline `=begin =end` comments inside hash
9. duplicate argument names now produce a parse error
10. defining a singleton method on a literal is not allowed anymore
11. `m <literal> {}` is not allowed anymore

### Motivation
By generating a test suite from `whitequark/parser` we can guarantee that all test cases are supported by the grammar.

Also `import_whitequark.rb` that is used to build these test files literally documents the difference between `whitequark/parser` and `sorbet/parser`. It rewrites AST for each test before emitting it, comments included.

### Test plan

`bazel test //test:whitequark_parser_tests --config=dbg` gives 0 errors, however I've changed the format of the AST a bit (e.g. no `dstr` nodes in a few places):

```
  expect
    Which is: "DString {\n  nodes = [\n    String {\n      val = <U \\xg3>\n    }\n  ]\n}\n"
  gotPhase.second
    Which is: "String {\n  val = <U \\xg3>\n}\n"
```

Because of that 10 tests fail from the `test_corpus`:
1. some of them produce a new "compact" version of the AST. should I manually update them?
2. `parser/anon_blockarg` - this guy is broken, I don't even know what is it. This syntax is invalid for all versions of Ruby. Should I remove it?
3. `desugar/fuzz_break_do_end` - no longer valid; rm?
4. `namer/constant_in_block` - no longer valid; rm?
5. `namer/dynamic_constant` - no longer valid; rm?
6. `namer/fuzz_repeated_kwarg` - no longer valid; rm?
7. `resolver/type_member_in_method` - no longer valid; rm?

Feel free to ask any questions, I backported some of these changes to `whitequark/parser`, so I can explain the rationale behind all of them.

### CODE REVIEW

**Comment by iliabylich** [comment]
Note for reviewers: this PR is not that big, there's a ton of auto-generated files in `test/testdata/parser/whitequark/*`. I didn't want to split it across multiple PRs because I wanted to be sure that both lexer and parser do the right thing. With all these test files I am sure that there are no bugs. Sorry about it.

**Comment by jez** [comment]
Heads up I'm looking at this right now, but it's taking a while. Thank you for your patience.

I tested this branch against our Ruby monorepo, and it turned up 20 new errors that all minimize to this example:

```ruby
# typed: strict

match ~A & B do
end
```

The output of `parse-tree-whitequark` differs in the node types that are created for `A` and `B`:

**Before this change:**

```
s(:block,
  s(:send, nil, :match,
    s(:send,
      s(:send,
        s(:const, nil, :A), :~), :&,
      s(:const, nil, :B))), nil, nil)
```

**After this change:**

```
s(:send, nil, :match,
  s(:send,
    s(:send,
      s(:const, nil, :A), :~), :&,
    s(:block,
      s(:send, nil, :B), nil, nil))
```

There are two differences:

- the `s(:const, nil, :B)` has become a `s(:send, nil, :B)`
- the `s(:block, ...)` binds to this newly created send, instead of the `s(:send, nil, :match)`

Please take a look. We can't merge this into Sorbet until this is fixed. In the mean time I'll keep reviewing your change.

**Comment by jez** [comment]
Upon further inspection, this is what the parse tree looks like from ripper:

```
irb(main):001:0> require 'ripper'
irb(main):002:0> require 'pp'
irb(main):003:0> pp Ripper.sexp('match ~A & B do; end')
[:program,
 [[:method_add_block,
   [:command,
    [:@ident, "match", [1, 0]],
    [:args_add_block,
     [[:binary,
       [:unary, :~, [:var_ref, [:@const, "A", [1, 7]]]],
       :&,
       [:var_ref, [:@const, "B", [1, 11]]]]],
     false]],
   [:do_block, nil, [[:void_stmt]]]]]]
```

**Comment by iliabylich** [comment]
@jez that's a bug in the whitequark/parser. I'll fix and backport it, thank you. This is the only error, right?

Als there are 20 failing specs on the CI (as I mentioned above) - what should I do? Just update them to match new AST format?

**Comment by iliabylich** [comment]
I've prepared a fix for whitequark/parser and backported it into this branch. It's not merged yet, but here's how it works with sorbet:
```
$ SORBET_SILENCE_DEV_MESSAGE=1 bazel-bin/main/sorbet --stop-after parser -p parse-tree-whitequark --no-error-count -e 'match ~A & B do; end'
s(:block,
  s(:send, nil, :match,
    s(:send,
      s(:send,
        s(:const, nil, :A), :~), :&,
      s(:const, nil, :B))), nil, nil)
```

Also I've fixed warnings in the bash script and updated test files to match new AST structure. The build on the CI should be green now (at least `bazel test //...` gives 0 errors locally and there were no errors from sanitizers on the CI)

**Comment by iliabylich** [comment]
@jez The PR has been merged into `whitequark/parser`. I've rebased my branch and regenerated the test suite to have a test for this bug.

Ready for reviewing, internal testing and (hopefully) merging

**Comment by jez** [comment]
@iliabylich i've re-posted this to our internal CI. i'll take a stab at reviewing it sometime later today.

**Comment by jez** [comment]
Thank you for the fix upstream to whitequark. It looks like it type checks with no new errors on Stripe's monorepo now.

File: parser/Builder.cc [hunk 1]
```diff
@@ -8,6 +8,7 @@
 #include "ruby_parser/diagnostic.hh"
 
 #include <algorithm>
+#include <map>
```
Comment (jez) [inline parser/Builder.cc]: Instead of std::map, can you rewrite this code to use `UnorderedMap`, which is a type alias that we've defined in `common.h`?

https://github.com/iliabylich/sorbet/blob/6499a58a7d8919a43f52d45a10a2d93af9d1d95e/common/common.h#L32

You should mostly be able to use the exact same APIs, but just will need to change the `include` and the method signatures.

File: test/testdata/namer/fuzz_repeated_kwarg.rb [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
-def f5(x, x:); end # error: Duplicated argument name `x`
-def f6(x, x: nil); end # error: Duplicated argument name `x`
-def f7(x=0, x: nil); end # error: Duplicated argument name `x`
+def f1(x, x); end # error: Parse Error: duplicate argument name
+def f2(x, x=0); end # error: Parse Error: duplicate argument name
+def f3(x:, x:); end # error: Parse Error: duplicate argument name
+def f4(x:, x: nil); end # error: Parse Error: duplicate argument name
+def f5(x, x:); end # error: Parse Error: duplicate argument name
+def f6(x, x: nil); end # error: Parse Error: duplicate argument name
+def f7(x=0, x: nil); end # error: Parse Error: duplicate argument name
```
Comment (jez) [inline test/testdata/namer/fuzz_repeated_kwarg.rb]: This is a regression in error message quality to me:

- The `Parse Error:` prefix is inconsistent with every other error message from Sorbet. (Sorbet doesn't say what pass the error came from, only what the error is.)
- The error message no longer mentions which argument is duplicated.

Can we fix either or both of these?

File: test/testdata/namer/dynamic_constant.rb [hunk 3]
```diff
@@ -3,6 +3,6 @@
 class A
   def self.foo
     X = 10
-  # ^^^^^^ error: Dynamic constant assignment
+#   ^ error: Parse Error: dynamic constant assignment
```
Comment (jez) [inline test/testdata/namer/dynamic_constant.rb]: ditto https://github.com/sorbet/sorbet/pull/1670/files#r322410162

Can we get rid of the `Parse Error:` prefix?

File: test/BUILD [hunk 4]
```diff
@@ -150,27 +150,57 @@ load(":pipeline_test.bzl", "pipeline_tests")
...
+    glob(
+        [
+            "testdata/**/*.rb",
+            "testdata/**/*.exp",
+            "testdata/**/*.exp.svg",
+        ],
+        exclude = [
+            "testdata/parser/whitequark/**/*.rb",
+            "testdata/parser/whitequark/**/*.exp",
+        ],
```
[the remark this answers is not in the document]
Comment (iliabylich) [reply test/BUILD]: I'd prefer the latter, they are special since we don't care about any steps except parsing.

File: test/test_corpus.cc [hunk 5]
```diff
@@ -480,6 +495,116 @@ TEST_P(ExpectationTest, PerPhaseTest) { // NOLINT
...
+    vector<unique_ptr<core::Error>> errors;
+    Expectations test = GetParam();
+    auto inputPath = test.folder + test.basename;
+    auto rbName = test.basename + ".rb";
+    SCOPED_TRACE(inputPath);
+
+    for (auto &exp : test.expectations) {
+        auto it = knownExpectations.find(exp.first);
+        if (it == knownExpectations.end()) {
+            ADD_FAILURE() << "Unknown pass: " << exp.first;
```
[the remark this answers is not in the document]
Comment (iliabylich) [reply test/test_corpus.cc]: huh, I didn't really know what this piece of code does, thanks, will do!

File: test/testdata/namer/fuzz_repeated_kwarg.rb (hunk 2 above)
Comment (iliabylich) [reply test/testdata/namer/fuzz_repeated_kwarg.rb]: well, it needs explanation I guess. In master Sorbet's parser doesn't emit any errors for this case, it parses it into an AST and so later some step generates an error for it. So now there are two errors:
```
$ SORBET_SILENCE_DEV_MESSAGE=1 bazel-bin/main/sorbet -p parse-tree-whitequark --no-error-count -e 'def m(a, a); end'
s(:def, :m,
  s(:args,
    s(:arg, :a),
    s(:arg, :a)), nil)
-e:1: Parse Error: duplicate argument name https://srb.help/2001
     1 |def m(a, a); end
                 ^

-e:1: Duplicated argument name a https://srb.help/4018
     1 |def m(a, a); end
                 ^
    -e:1: In argument list of method
     1 |def m(a, a); end
        ^^^^^^^^^^^
```

Technically it's a parsing error, should I remove the code that gives an error after parsing?

> The `Parse Error:` prefix is inconsistent with every other error message from Sorbet. (Sorbet doesn't say what pass the error came from, only what the error is.)
That's a prefix that every parse error has. Should I remove it?

> The error message no longer mentions which argument is duplicated.
Yes, that's what what whitequark/parser returns. Not sure if I can change but I'll give it a try. And I'll try to change an `unexpected token` error as well to include the token name.

File: test/testdata/parser/misc.rb.parse-tree.exp [hunk 6]
```diff
@@ -341,7 +337,7 @@ Begin {
       body = Nil {
       }
     }
-    WhilePost {
+    While {
```
[the remark this answers is not in the document]
Comment (iliabylich) [reply test/testdata/parser/misc.rb.parse-tree.exp]: I mentioned it in the beginning: `foo while bar` is not a post-while loop. That's a bug in the master branch. Same for `foo until bar`

File: test/test_corpus.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (iliabylich) [reply test/test_corpus.cc]: hm, this what it does atm:
``` sh
$ echo '123' >  test/whitequark/test_alias_0.parse-tree-json.exp
$ bazel test //test:test_WhitequarkParserTests/whitequark/test_alias_0
INFO: Invocation ID: fc4ef863-c846-4fea-89a4-6651ebd04993
INFO: Analysed target //test:test_WhitequarkParserTests/whitequark/test_alias_0 (0 packages loaded, 0 targets configured).
INFO: Found 1 test target...
Target //test:test_WhitequarkParserTests/whitequark/test_alias_0 up-to-date:
  bazel-bin/test/test_WhitequarkParserTests/whitequark/test_alias_0
INFO: Elapsed time: 0,327s, Critical Path: 0,01s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
//test:test_WhitequarkParserTests/whitequark/test_alias_0       (cached) PASSED in 0.5s

INFO: Build completed successfully, 1 total action
```
i.e. it ignores all other expectation files. Could you explain what's missing please?

File: core/errors/namer.h [hunk 7]
```diff
@@ -20,7 +20,6 @@ constexpr ErrorClass DynamicConstant{4014, StrictLevel::False};
 constexpr ErrorClass InvalidClassOwner{4015, StrictLevel::False};
 constexpr ErrorClass RootTypeMember{4016, StrictLevel::False};
 constexpr ErrorClass DynamicConstantAssignment{4017, StrictLevel::False};
-constexpr ErrorClass RepeatedArgument{4018, StrictLevel::False};
```
Comment (DarkDimius) [inline core/errors/namer.h]: please, to remove errors, comment them out. This allows us to not reuse error codes.

File: test/BUILD [hunk 8]
```diff
@@ -159,6 +159,17 @@ pipeline_tests(
     filter = "PosTests",
 )
 
+pipeline_tests(
```
Comment (DarkDimius) [inline test/BUILD]: these tests are already included in `test_corpus`, right?
why would we register & run them twice?

File: test/testdata/parser/whitequark/test_ruby_bug_11873_a_16.rb [hunk 9]
```diff
@@ -0,0 +1,3 @@
+# typed: true
+
+a b{c d}, 1.0i do end
```
Comment (DarkDimius) [inline test/testdata/parser/whitequark/test_ruby_bug_11873_a_16.rb]: why do these files sometimes have end of line and sometimes don't?

File: test/testdata/parser/whitequark/assert_parses_args.rb [hunk 10]
```diff
@@ -1,3 +0,0 @@
-# typed: true
-
```
Comment (DarkDimius) [inline test/testdata/parser/whitequark/assert_parses_args.rb]: was this test removed?

File: test/testdata/parser/whitequark/assert_parses_args_10.rb [hunk 11]
```diff
@@ -1,4 +0,0 @@
-# typed: true
-
-def f foo:
```
Comment (DarkDimius) [inline test/testdata/parser/whitequark/assert_parses_args_10.rb]: ditto

File: test/test_corpus.cc [hunk 12]
```diff
@@ -569,8 +569,8 @@ TEST_P(WhitequarkParserTest, PerPhaseTest) { // NOLINT
         {
             core::UnfreezeNameTable nameTableAccess(gs); // enters original strings
 
-            nodes = parser::Parser::run(gs, file);
-            TEST_COUT << "PARSED NODE" << nodes->nodeName() << "\n";
+            vector<string> locals = {"foo", "bar", "baz"};
+            nodes = parser::Parser::runWithLocals(gs, file, locals);
```
Comment (DarkDimius) [inline test/test_corpus.cc]: Could you give more context to this change?
I'd find it super surprising that the commonly used function names have already been used for local variables.

File: parser/Builder.cc [hunk 13]
```diff
@@ -940,7 +940,17 @@ class Builder::Impl {
     }
 
     unique_ptr<Node> symbol_compose(const token *begin, sorbet::parser::NodeVec parts, const token *end) {
-        return make_unique<DSymbol>(collectionLoc(begin, parts, end), std::move(parts));
+        if (collapseStringParts(parts)) {
```
Comment (DarkDimius) [inline parser/Builder.cc]: I didn't understand motivation for this change

File: parser/Builder.cc [hunk 14]
```diff
@@ -8,6 +8,7 @@
 #include "ruby_parser/diagnostic.hh"
 
 #include <algorithm>
+#include <iostream>
```
Comment (DarkDimius) [inline parser/Builder.cc]: please remove this

**Comment by DarkDimius** [comment]
Hi Ilya,

thank you for your effort, appreciate it a lot!
Would you be up to work together to split this PR into multiple?
There are multiple things that being changed in this PR which make it both hard to review _and_ will make following the commit history hard in future. This will also make it possible for us to bisect code in case of regressions.

I'd suggest splitting up each idependent change in PR of its own. Would you be up for that? Happy to help.
Again, super appreciate your effort and happy to pair with you to get this in.

File: test/testdata/parser/whitequark/assert_parses_args_10.rb (hunk 11 above)
Comment (iliabylich) [reply test/testdata/parser/whitequark/assert_parses_args_10.rb]: this guy moved to `kwarg_no_paren`. `whitequark/parser` has a few utility methods for testing and `assert_parses_args` is one of them. The name of the real test is `test_kwarg_no_paren`

File: test/testdata/parser/whitequark/assert_parses_args.rb (hunk 10 above)
Comment (iliabylich) [reply test/testdata/parser/whitequark/assert_parses_args.rb]: it's `test_marg_combinations_0` now

File: test/testdata/parser/whitequark/test_ruby_bug_11873_a_16.rb (hunk 9 above)
Comment (iliabylich) [reply test/testdata/parser/whitequark/test_ruby_bug_11873_a_16.rb]: by EOL you mean an empty trailing newline, right? If so all files have it:
```
        File.write(input_filepath, "# typed: true\n\n" + capture[:input] + "\n")

        if capture[:output]
          File.write(output_filepath, capture[:output] + "\n")
        end
```

File: test/test_corpus.cc (hunk 12 above)
Comment (iliabylich) [reply test/test_corpus.cc]: basically `whitequark/parser` introduces a few local variables before running each test: https://github.com/whitequark/parser/blob/master/test/test_parser.rb#L16-L18. It simplifies a lot of common use cases when you want to parse something that uses lvar without declaring it explicitly

File: parser/Builder.cc (hunk 13 above)
Comment (iliabylich) [reply parser/Builder.cc]: `whitequark/parser` emits a string `"foo"` as `s(:str, "foo")`, but sorbet/parser emits `s(:dstr, s(:str, "foo"))`. Same for `xstr` by the way.

This `dstr` node doesn't have any useful information in its source maps (because there's simply no additional context that you could add, `s(:str)` has everything you need).

But I ported it to match `whitequark/parser`'s structure, probably I could do that via adding a few more logic to the rewriter, but I think it's not worth it.

File: parser/Builder.cc (hunk 14 above)
Comment (iliabylich) [reply parser/Builder.cc]: This line is not there anymore, looks like your review is a bit outdated.

**Comment by DarkDimius** [comment]
Test corpus is expected to mimic behaviour of normal typechecking as 
closely as possible.

The intention is that every user should be able to easily add an rb file 
and there should be no details that need to have in mind.

As this is used not only to test parser, but to test all of Sorbet, I dont 
think if we deviate from Ruby in tests by silently introducing local variables

**Comment by iliabylich** [comment]
> As this is used not only to test parser, but to test all of Sorbet, I dont
think if we deviate from Ruby in tests by silently introducing local variables

I thought this test runner (`test_corpus --gtest_filter=WhitequarkParserTests/*`) is used only for files under `whitequark/parser` directory. All other tests are going through existing `ExpectationTest/LSPTest`

> I'd suggest splitting up each independent change in PR of its own

I agree, WDYT about the following sequence of PRs:
1. add `import_whitequark.[sh|rb]` + test runner
2. import test files from `whitequark/parser` using `import_whitequark.sh`
3. bazel task that runs `test/whitequark` (using a runner from item 1) + some basic changes to match the structure (like the change to `dstr/xstr`) + exclude list for everything that doesn't pass
4. fix the rest and cleanup an 'exclude' filter

item 3 could be split into 2 more steps:
1. filter everything that doesn't pass (there will be a lot of files)
2. fix `str/dstr` + add them back to the test suite

**Comment by DarkDimius** [comment]
> I thought this test runner (test_corpus --gtest_filter=WhitequarkParserTests/*) is used only for files under whitequark/parser directory. All other tests are going through existing ExpectationTest/LSPTest

Oh, this changes things dramatically. Agreed than! Sorry, it was hard to follow for me.
Let's split it up as you suggest, with splitting steps 3 and 4 where it makes sense.

**Comment by iliabylich** [comment]
@DarkDimius what does CI run? `bazel test //...`? Just to be sure that I'm not sending broken intermediate PRs with all these temporary filters. I understand that CI will run them anyway, just want to have some level of confidence before submitting them

**Comment by iliabylich** [comment]
Finally `+894 −637`!

To sum up:
+ this PR doesn't include scripts that import tests (https://github.com/sorbet/sorbet/pull/1875)
+ it doesn't include tests (#1895)
+ it doesn't include improvements to parse errors because it affected too many tests (https://github.com/sorbet/sorbet/pull/1876)
+ it doesn't include update to `dstr`/`dsym` nodes for the same reason (https://github.com/sorbet/sorbet/pull/1894)

It still includes:
+ `foo while bar` is no longer a post-while loop (`raise while false` doesn't raise, it's just a regular while loop), same for `foo unless bar`
+ defining constants/modules/dynamic constants in methods is no longer allowed
+ fixed parsing of `def BEGIN; end`
+ `lex.in_kwarg` is used now (I think typedruby removed it intentionally), so the code like `def a b:\nreturn\nend` can be parsed now.
+ introduced `ruby_parser::Context` (similar to `Parser::Context` in `whitequark/parser`)
+ fixed parsing of multiline `=begin =end` comments inside hash
+ duplicate argument names now produce a parse error
+ defining a singleton method on a literal is not allowed anymore
+ `m <literal> {}` is not allowed anymore, that's a syntax error

@DarkDimius / @jez PTAL when you have time

**Comment by jez** [comment]
Thank you for the summary, and sorry for the delay! I'll get back to you on this today.

**Comment by iliabylich** [comment]
@jez Thanks! I need some time to rebase with https://github.com/sorbet/sorbet/pull/1993

**Comment by jez** [comment]
(sorry, accidentally hit "request changes" instead of "approve")

**Comment by iliabylich** [comment]
@jez Done, I had to simplify a few tests because of the error recovery mechanism. It gives wrong errors (and `whitequark/parser` also doesn't handle it well) - https://sorbet.run/#def%20m(a%20%3D%20def%20n(A)%3B%20end)%3B%20end - and there are many more. Some of them are much more complicated to reproduce
